spock-spring/src/main/java/org/spockframework/spring/SpringExtension.java

```diff
@@ -93,6 +96,51 @@ private boolean isSpringSpecUsingFindAnnotationDescriptorForTypes(SpecInfo spec)
         new Class[] {ContextConfiguration.class, contextHierarchyClass, bootstrapWithAnnotation}) != null;
   }
 
+  private void verifySharedFieldsInjection(SpecInfo spec) {
+    if (spec.isAnnotationPresent(EnableSharedInjection.class)) {
+      verifySharedFieldsInjectionEnabled(spec);
+    } else {
+      checkNoSharedFieldsInjected(spec);
+    }
+  }
+
+  private void verifySharedFieldsInjectionEnabled(SpecInfo spec) {
+    if (spec.isAnnotationPresent(DirtiesContext.class)) {
+      ClassMode classMode = spec.getAnnotation(DirtiesContext.class).classMode();
+      if (classMode == ClassMode.BEFORE_EACH_TEST_METHOD || classMode == ClassMode.AFTER_EACH_TEST_METHOD) {
+        throw sharedInjectionWithDirtiesContextException();
+      }
+    }
+    if (isTransactionalAnnotationPresent(spec)) {
```

Comment:
And how would we
> have a warning in the annotation present
?

I would assume you mean to log that somehow but I could not find a precedent of logging or warning about anything anywhere in the codebase so I don't know how to address this. Can you please advise?

Reply:
I mean just mention it in the Javadoc, in a little more detail. Something akin to:

> Using shared injection together with `@Transactional` can lead to unexpected results as the shared initializer and `setupSpec` runs in a different transaction than the individual features.

We actually would need to test if that is the case, I'm not sure that it even would run in a transaction at all. The same with the `where` block, I think it might be executed outside of a transaction as well.

Reply:
I see. Makes sense and is now actionable for me, thanks.

Reply:
Have you actually verified the behavior of `@Transactional` with `where:` and `setupSpec`, as I don't see any tests regarding that.

Reply:
No, I did not explicitly verify that behaviour, I just assumed that transactions will be started when `org.spockframework.spring.SpringTestContextManager.beforeTestMethod()` is called. 

Now that you asked I actually verified that - I put a breakpoint in `org.springframework.jdbc.datasource.DataSourceTransactionManager.doBegin()` and then debugged `org.spockframework.spring.TransactionalExample`. When the breakpoint was hit `org.spockframework.spring.SpringTestContextManager.beforeTestMethod()` was indeed in the stacktrace. It also helped me to establish that transaction support in spring-test is handled by `org.springframework.test.context.transaction.TransactionalTestExecutionListener`. If you inspect that class you will see that both `beforeTestClass()` and `prepareTestInstance()` for that class have empty implementations and these two methods are the only ones called before `setupSpec()` - see `org.spockframework.spring.SpringInterceptor.interceptSetupSpecMethod()`. 

Secondly, as far as I understand `where:` blocks are being executed before feature method bodies are executed which means that `where:` blocks are executed before `org.spockframework.spring.SpringInterceptor.interceptSetupMethod()` which starts the transaction via `org.spockframework.spring.SpringTestContextManager.beforeTestMethod()`.